Add `make help` command to list available Makefile targets and descriptions

Naman-B-Parlecha opened the pull request:
<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

Resolves: #16729 

This PR introduces a `make help` command that displays all available Makefile targets along with their descriptions to eliminating the need to manually scan through the Makefile.

Usage: 
```bash
make help
```

Result: 

```
Usage:
  make <target>

Main Targets:
  help                  Displays commands and their descriptions.
  update-npm-deps       Update Prometheus web UI npm dependencies to minor versions
  upgrade-npm-deps      Upgrade Prometheus web UI npm dependencies to latest versions
  ui-bump-version       Bump Prometheus web UI version and update package files
  ui-install            Install npm dependencies for Prometheus web UI
  ui-build              Build Prometheus web UI assets
  ui-build-module       Build Prometheus web UI as a module
  ui-test               Run Prometheus web UI tests
  ui-lint               Lint Prometheus web UI code
  assets                Build and embed Prometheus web UI assets (Skipped if using pre-built assets)
  npm_licenses          Bundle npm license files for Prometheus web UI (Skipped if using pre-built assets)
  assets-compress       Compress Prometheus web UI assets for distribution
  assets-tarball        Package Prometheus web UI assets into tarball
  parser                Generate Prometheus PromQL parser Go code using goyacc
  clean-parser          Remove generated Prometheus PromQL parser files
  check-generated-parser  Verify Prometheus PromQL parser is up to date
  install-goyacc        Install goyacc tool for Prometheus PromQL parser generation
  test                  Run all Prometheus tests including Go, UI, and parser checks (To test Go only set GO_ONLY=1)
  tarball               Create Prometheus release tarball with licenses
  docker                Build Prometheus Docker image with licenses
  plugins               Generate Prometheus plugins list from plugins.yml
  build                 Build complete Prometheus binary with UI, licenses, and plugins
  bench_tsdb            Run Prometheus TSDB benchmarks and generate profiling reports
  cli-documentation     Generate Prometheus and promtool command-line documentation
  check-go-mod-version  Verify Prometheus go.mod version consistency
  update-all-go-deps    Update all Prometheus Go dependencies including examples

Common Targets:
  common-all            Run all Prometheus quality checks and build
  common-style          Check Go code formatting with gofmt
  common-check_license  Verify Prometheus license headers in Go files
  common-deps           Download Go module dependencies
  update-go-deps        Update all Go dependencies to latest versions
  common-test-short     Run short Go tests only
  common-test           Run all Go tests including race detection
  common-format         Format Go code with gofmt and golangci-lint
  common-vet            Run Go vet static analysis
  common-lint           Run golangci-lint code analysis
  common-lint-fix       Run golangci-lint with auto-fix enabled
  common-yamllint       Lint all YAML files in repository
  common-staticcheck    Run static code analysis (deprecated, use lint)
  common-unused         Check for unused dependencies in go.mod
  common-build          Build Prometheus project binaries using promu
  common-tarball        Create Prometheus release tarball with binaries
  common-docker-repo-name  Print Prometheus Docker repository name
  common-docker         Build Prometheus Docker images for all architectures
  common-docker-publish  Push Prometheus Docker images to registry
  common-docker-tag-latest  Tag Prometheus Docker images as latest and version
  common-docker-manifest  Create and push multi-arch Prometheus Docker manifest
  promu                 Install or ensure promu Prometheus build tool is available
  common-proto          Generate Go code from protobuf files
  precheck              Run prerequisite checks for required Prometheus tools
  govulncheck           Run Go vulnerability scanner
  install-govulncheck   Install Go vulnerability checker tool
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-notes block below.
Otherwise, please describe what should be mentioned in the CHANGELOG. Use the following prefixes:
[FEATURE] [ENHANCEMENT] [PERF] [BUGFIX] [SECURITY] [CHANGE]
Refer to the existing CHANGELOG for inspiration:  https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md
If you need help formulating your entries, consult the reviewer(s).
-->
```release-notes
NONE
```

bwplotka commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 ## 3.4.1 / 2025-05-31
 
 * [BUGFIX] Parser: Add reproducer for a dangling-reference issue in parsers. #16633
+* [ENHANCEMENT] Add `make help` command to list available Makefile targets and descriptions #16731
```
Let's remove it, while this change is useful it's not surfacing users.

bwplotka commented on Makefile:
```diff
@@ -35,44 +35,50 @@ ifdef PREBUILT_ASSETS_STATIC_DIR
   SKIP_UI_BUILD = true
 endif
 
+help: ## Displays commands and their descriptions.
```
This is great, I only wonder if we shouldn't do this in `Makefile.common` which we propagate cross organization (via https://github.com/prometheus/prometheus/blob/main/scripts/sync_repo_files.sh).

WDYT @SuperQ ?

bwplotka commented on Makefile:
```diff
@@ -175,20 +181,20 @@ bench_tsdb: $(PROMU)
 
 .PHONY: update-all-go-deps
-update-all-go-deps:
+update-all-go-deps: ## Update all Prometheus Go dependencies including examples
```
personally I would use full sentence (with trailing period), but not a blocker.
```suggestion
update-all-go-deps: ## Update all Prometheus Go dependencies including examples.
```

bwplotka commented on Makefile:
```diff
@@ -175,20 +181,20 @@ bench_tsdb: $(PROMU)
 		$(GO) get -d $$m; \
 	done
-	@cd ./documentation/examples/remote_storage/ && $(GO) mod tidy
+	@cd ./documentation/examples/remote_storage/ && $(GO) mod tidy
```
Why newline remove? (and below)

bwplotka commented on Makefile.common:
```diff
@@ -107,10 +107,10 @@ endif
 
 .PHONY: common-all
-common-all: precheck style check_license lint yamllint unused build test
+common-all: precheck style check_license lint yamllint unused build test ## Run all Prometheus quality checks and build
```
For long lines I would suggest creating extra line like here:
 https://github.com/thanos-io/thanos/blob/main/Makefile#L228C1-L228C52
```suggestion
common-all:  ## Run all Prometheus quality checks and build.
common-all: precheck style check_license lint yamllint unused build test
```

bwplotka commented on documentation/examples/Makefile:
```diff
@@ -11,10 +11,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+help: ## Displays commands and their descriptions.
```
include Makefile.common?

Naman-B-Parlecha replied:
included

Naman-B-Parlecha replied:
so just including the Makefile.common in all the sub dir should work

bwplotka commented on documentation/examples/remote_storage/Makefile:
```diff
@@ -14,7 +14,11 @@
 include ../../../Makefile.common
 
-build:
+help: ## Display available commands and their descriptions
```
ditto

bwplotka replied:
Sorry, no, I mean https://github.com/prometheus/prometheus/pull/16732/files#r2158272954

bwplotka commented on documentation/prometheus-mixin/Makefile:
```diff
@@ -1,19 +1,22 @@
 JSONNET_FMT := jsonnetfmt -n 2 --max-blank-lines 2 --string-style s --comment-style s
 
-all: fmt prometheus_alerts.yaml dashboards_out lint
+help: ## Display available commands and their descriptions
```
ditto

Naman-B-Parlecha replied:
same doubt here

bwplotka replied:
Sorry, I mean https://github.com/prometheus/prometheus/pull/16732/files#r2158272954

bwplotka reviewed (commented):
Nice!

Some comments and suggestion, but overall looks good. Thanks!

I do think some basic help and nice list of possible commands can help 💪🏽

Naman-B-Parlecha commented:
@bwplotka refactored, have a look and let me know if any changes are required

bwplotka commented on documentation/prometheus-mixin/Makefile:
```diff
@@ -22,8 +29,8 @@ lint: prometheus_alerts.yaml
-clean:
-	rm -rf dashboards_out prometheus_alerts.yaml
+clean: ## Remove generated files (dashboards and alert rules).
+	rm -rf dashboards_out prometheus_alerts.yaml
```
Please add back newlines at the end (:

Naman-B-Parlecha replied:
added back in

bboreham commented on Makefile.common:
```diff
@@ -106,11 +106,16 @@ endif
 
+.PHONY: common-help
+common-help: ## Display available commands and their descriptions.
+	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[38;2;230;82;44m<target>\033[0m\n\nTargets:\n"} /^[a-z0-9A-Z_-]+:.*?##/ { printf "  \033[38;2;230;82;44m%-20s\033[0m %s\n", $$1, $$2 }' $(MAKEFILE_LIST)
```
Why do we need an `awk` command in both `Makefile` and `Makefile.common` ?

bboreham commented on documentation/prometheus-mixin/Makefile:
```diff
@@ -1,19 +1,28 @@
 JSONNET_FMT := jsonnetfmt -n 2 --max-blank-lines 2 --string-style s --comment-style s
 
+include ../../Makefile.common
```
This might not be correct - the top-level `Makefile.common` is for Prometheus components and not designed for mixins which have their own home at https://monitoring.mixins.dev/.

bboreham reviewed (commented):
Hello from the bug-scrub!
Thanks for this, seems generally good but I'm unclear whether the split between `Makefile` and `Makefile.common` is correct.

Naman-B-Parlecha commented:
Hey @bwplotka @bboreham, I have refactored the code abit, I have removed the `help` from `Makefile.common` and will no longer be including `Makefile.common` in any sub `Makefile` (eg /documentation/examples/Makefile) - [reference](https://github.com/prometheus/prometheus/pull/16732#discussion_r2158272954) 

We will explicitly have `help` in each `Makefile` initially for prometheus repo since `Makefile.common` would propagate to cross organization project

We can start by introducing this for prometheus repo initially :)
